examples/cis2-wccd-upgradable/src/lib.rs

```diff
@@ -0,0 +1,3721 @@
+fn contract_state_get_paused<S: HasStateApi>(
+    _ctx: &impl HasReceiveContext,
+    host: &impl HasHost<State<S>, StateApiType = S>,
+) -> ContractResult<bool> {
+    Ok(host.state().paused)
+}
+
+/// Get Balance.
+#[receive(
+    contract = "CIS2-wCCD-State",
+    name = "getBalance",
+    parameter = "GetBalanceParams",
+    return_value = "ContractTokenAmount"
+)]
+fn contract_state_get_balance<S: HasStateApi>(
+    ctx: &impl HasReceiveContext,
+    host: &impl HasHost<State<S>, StateApiType = S>,
+) -> ContractResult<ContractTokenAmount> {
+    let params: GetBalanceParams = ctx.parameter_cursor().get()?;
+
+    if let Some(s) = host.state().token.get(&params.owner) {
+        Ok(s.balance)
+    } else {
+        Ok(0u64.into())
+    }
+}
+
+/// Get Implementors.
+#[receive(
+    contract = "CIS2-wCCD-State",
+    name = "getImplementors",
+    parameter = "GetImplementorsParams",
+    return_value = "ContractTokenAmount"
+)]
+fn contract_state_get_implementors<S: HasStateApi>(
+    ctx: &impl HasReceiveContext,
+    host: &impl HasHost<State<S>, StateApiType = S>,
+) -> ContractResult<Option<Vec<ContractAddress>>> {
+    let params: GetImplementorsParams = ctx.parameter_cursor().get()?;
+
+    if let Some(addresses) = host.state().implementors.get(&params.id) {
+        let return_addresses = &*addresses;
+        Ok(Some(return_addresses.to_vec()))
+    } else {
+        Ok(None)
+    }
+}
+
+/// Is_operator.
+#[receive(
+    contract = "CIS2-wCCD-State",
+    name = "isOperator",
+    parameter = "IsOperatorParams",
+    return_value = "bool"
+)]
+fn contract_state_get_operator<S: HasStateApi>(
+    ctx: &impl HasReceiveContext,
+    host: &impl HasHost<State<S>, StateApiType = S>,
+) -> ContractResult<bool> {
+    let params: IsOperatorParams = ctx.parameter_cursor().get()?;
+
+    Ok(host
+        .state()
+        .token
+        .get(&params.owner)
+        .map_or(false, |address_state| address_state.operators.contains(&params.address)))
+}
+
+/// Function to receive CCD on the implementation. All CCDs are held by the
+/// proxy contract. Nonetheless, during the `unwrap` function the proxy should
+/// send some CCD amount back to the implementation which then goes through
+/// additional logic to deal with sending the CCD to the receiver address.
+#[receive(contract = "CIS2-wCCD", name = "receiveCCD", payable)]
+fn contract_implementation_recieve_ccd<S: HasStateApi>(
+    _ctx: &impl HasReceiveContext,
+    _host: &impl HasHost<StateImplementation, StateApiType = S>,
+    _amount: Amount,
+) -> ContractResult<()> {
+    Ok(())
+}
+
+/// Function to receive CCD on the proxy. All CCDs are held by the
+/// proxy contract.
+#[receive(contract = "CIS2-wCCD-Proxy", name = "receiveCCD", payable)]
+fn contract_proxy_recieve_ccd<S: HasStateApi>(
+    _ctx: &impl HasReceiveContext,
+    _host: &impl HasHost<StateProxy, StateApiType = S>,
+    _amount: Amount,
+) -> ContractResult<()> {
+    Ok(())
```

Comment:
It would probably be a good idea to fail when not called by the implementation contract.
Since we want a 1:1 with wCCD and the amount of CCD in this contract.

Reply:
Added restrictions that the two `receiveCCD` functions can only be called by `implementation` or `proxy` respectively.